xbmc/platform/linux/input/LibInputHandler.cpp

```diff
@@ -0,0 +1,249 @@
+  }
+}
+
+void CLibInputHandler::ProcessEvent(libinput_event *ev)
+{
+  libinput_event_type type = libinput_event_get_type(ev);
+  libinput_device *dev = libinput_event_get_device(ev);
+
+  switch (type)
+  {
+    case LIBINPUT_EVENT_DEVICE_ADDED:
+    {
+      const char *sysname = libinput_device_get_sysname(dev);
+      const char *name = libinput_device_get_name(dev);
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_POINTER))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - pointer type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_KEYBOARD))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device == m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - keyboard type device added: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.push_back(dev);
+        }
+        m_keyboard->GetRepeat(dev);
+      }
+      break;
+    }
+    case LIBINPUT_EVENT_DEVICE_REMOVED:
+    {
+      const char *sysname = libinput_device_get_sysname(dev);
+      const char *name = libinput_device_get_name(dev);
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device != m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device removed: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.erase(device);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_POINTER))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device != m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - pointer type device removed: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.erase(device);
+        }
+      }
+
+      if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_KEYBOARD))
+      {
+        auto device = std::find(m_devices.begin(), m_devices.end(), dev);
+        if (device != m_devices.end())
+        {
+          CLog::Log(LOGDEBUG, "CLibInputHandler::%s - keyboard type device removed: %s (%s)", __FUNCTION__, name, sysname);
+          m_devices.erase(device);
+        }
+      }
+      break;
+    }
```

Comment:
general style comment - it seems weird to have that long block of code followed by the below handling of calling helper functions. why not use helpers for added/removed?